django/thunderstore/repository/views/repository.py

```diff
@@ -400,12 +402,26 @@ def get_package_listing_or_404(
 class PackageDetailView(DetailView):
     model = PackageListing
 
+    def get(self, *args, **kwargs):
+        try:
+            return super().get(*args, **kwargs)
+        except RedirectListingException as exc:
```

Comment:
While this works, using exceptions for logic flow like this is usually bad practice.

How about something like:
```python
from functools import cache

class PackageDetailView(DetailView):
    model = PackageListing

    def get(self, *args, **kwargs):
        listing = self.get_object(*args, **kwargs)
        if listing.community != self.request.community:
            return redirect(exc.listing.get_full_url(), permanent=False)
        else:
            return super().get(*args, **kwargs)
    
    @cache
    def get_object(self, *args, **kwargs):
        try:
            listing = get_listing(
                owner=self.kwargs["owner"],
                name=self.kwargs["name"],
                community=self.request.community,
            )
        except Http404:  # Try to find in another community
            listing = get_listing(
                owner=self.kwargs["owner"],
                name=self.kwargs["name"],
            )
        if not listing.can_be_viewed_by_user(self.request.user):
            raise Http404("Package is waiting for approval or has been rejected")
        return listing
```

So changes here are:
1. Use the [functools.cache](https://docs.python.org/3/library/functools.html#functools.cache) decorator on `get_object` to avoid querying the object twice
2. Instead of using exceptions to control the logic flow, moved the if check into the `get` method of view view.

In case our python version is too old to contain the `functools.cache` decorator, you can use `functools.lru_cache(maxsize=None)` instead.

Might also want to verify that the cache works as intended (as the cache key will be based on `*args` and `**kwargs`, which means our `self.get_object` call parameters must match whatever Django does by default in order for them to get the same cache key. This could use an unit test that checks query counts when calling this view.)

Reply:
Also applies to the other view which uses exception based logic flow